File: solutions/scala/grade-school/src/main/scala/GradeSchool.scala
import scala.collection.immutable.SortedMap

class School {
  private type DB = Map[Int, Seq[String]]
  private var database: DB = SortedMap.empty

  def add(name: String, g: Int): Unit = {
    database += (g -> (grade(g) :+ name))
  }

  def db: DB = database

  def grade(g: Int): Seq[String] = {
    db.getOrElse(g, Seq.empty)
  }

  def sorted: DB = database.view.mapValues(_.sorted).toMap
}

File: solutions/python/grade-school/grade_school.py
class School:
    def __init__(self):
        self.db = {}
        self.status = []

    def add_student(self, name: str, grade: int):
        not_exist = name not in self.db
        self.status.append(not_exist)
        if not_exist:
            self.db[name] = grade

    def roster(self) -> list[str]:
        return [
            name for name, grade in sorted(
                self.db.items(),
                key=lambda item: (item[1], item[0])
            )
        ]

    def grade(self, grade_number: int) -> list[str]:
        return sorted([
            name for name, grade in self.db.items()
            if grade == grade_number
        ])

    def added(self) -> list[bool]:
        return self.status
